### ISSUE TITLE
How to change to isolationlevel of transactions

### ISSUE THREAD

**Issue by Charles89** [MarimerLLC/cslaforum#220]
I have a large OLTP database which is also used for reporting facilities. The reporting facilities are a core part of the same application that I created for editing data. The default isolationlevel seems to be read committed (both Sql Server and Oracle) and causes readers to wait for writers and writers to wait for readers. I want to improve performance by choosing a different isolationlevel for the reporting facilities to reduce the locking that currently takes place. Problem is I can't find a way to change the isolationlevel of the transaction.

`using (var ctx = TransactionManager<UniConnection, UniTransaction>.GetManager(this.ConnectionString, false))
{
  var currentIsolationLevel = ctx.Transaction.IsolationLevel; // readonly property
`

I've tried to workaround it by executing SET TRANSACTION ISOLATION LEVEL SNAPSHOT, but that doesn't do anything. Any pointers on how to change the isolationlevel?

**Comment by jonnybee** [comment]
Assuming that you are using CSLA then the easiest way is to use the TransactionalAttribute on the DataPortal_XYZ methods to set as Manual and then use the TransactionScope manually. 


    [Transactional(TransactionalTypes.Manual)]
    protected override void DataPortal_Fetch()
    {
                using (var sc = new TransactionScope(
                    TransactionScopeOption.RequiresNew,
                    new TransactionOptions
                        {
                            IsolationLevel = IsolationLevel.Snapshot,
                        }))
                {
                     // code for fetch inside here ... 
                }
    }

**Comment by Charles89** [comment]
Correct, I'm using CSLA and the DAL-code is in a separate factory class. Unfortunately applying your code to my environment doesn't have the desired effect and the reason probably is that I use the CSLA TransactionManager class. It always seems to be using the ReadCommited isolationlevel. I can get it to work by doing this

```
using (var ctx = ConnectionManager<UniConnection>.GetManager(this.ConnectionString, false))
{
    var transaction = ctx.Connection.BeginTransaction(System.Data.IsolationLevel.Snapshot);

    // code for fetch

    transaction.Commit();
}
```

but I really like the CSLA TransactionManager class taking care of the Commit and Rollback for me. Is there any chance for me to overrule the default IsolationLevel in certain scenarios?

```
using (var ctx = TransactionManager<UniConnection, UniTransaction>(this.ConnectionString, false))
{
    // code for fetch
}
```

**Comment by rockfordlhotka** [comment]
The challenge with this (good) idea is that every location that calls the `GetManager` method must be identical - the code calling that method can not assume it is creating the transaction, because it might be getting an existing one. You can't assume that any given call is at the root level.

We could add a parameter to `GetManager` where you could set the isolation level - but you'd always have to set it to the same thing everywhere, or you'd lose the ability to compose objects.

Or we could have a static that you set on app startup so _all transactions globally in your app_ would use a different isolation level. I don't know how useful that would be?

**Comment by Charles89** [comment]
Setting a static property to change all transactions globally is not the direction I was looking for, because I need two different isolationlevels: one for editing that uses the default isolationlevel of _readcommitted_ and one for reporting that uses _snapshot_. Besides in Sql Server it is already possible to globally 'change' the isolation level from _readcommitted_ to _snapshot_ using this script:

```sql
ALTER DATABASE MyDatabase SET ALLOW_SNAPSHOT_ISOLATION ON;
ALTER DATABASE MyDatabase SET READ_COMMITTED_SNAPSHOT ON;
```

This allows queries that run in a readcommitted transaction to access the versioned rows. Unfortunately I'm dealing with Oracle too so this script won't work there.

Adding an islolationlevel to the TransactionManager seems logical, but the inner workings and any side-effects this change may have, are unclear to me. I know that currently, if the parameters stay the same, the TransactionManager doesn't really create a new transaction. Instead it re-uses the current one and increments a counter. When the counter hits 0 it would either commit or rollback, but with this new parameter that logic needs to be extended. The counter should be counting per transaction with a different isolationlevel and either commit or rollback at the end within a certain isolationlevel. Would that make sense?

**Comment by rockfordlhotka** [comment]
The problem is in your calling code. So in your calling code you want to set the isolation level, which you can only do in the case that the transaction was just created. But you can't tell if it was just created or not because that fact is abstracted away by the TransactionManager.

I could let you set the value, and if it is too late I could throw an exception. But I suspect this would lead to some very unpredictable code - depending on your call chain sometimes it would work and sometimes it would throw. That's clearly undesirable.

Silent failure is probably worse, because you'll _think_ you have one isolation mode when you really don't.

I really don't see a way to make this work I'm afraid...

**Comment by Charles89** [comment]
Maybe you misunderstood what I was trying to say: adding an property or method to the Transactionmanager class so you can change it from the outside, wasn't my intention. What I was referring to was, as you suggested before, to pass an IsolationLevel to the constructor of the TransactionManager class. Seeing the IDbConnection has a BeginTransaction overload that accepts IsolationLevel as a parameter, isn't it possible to use that in TransactionManager? For example:

```csharp
using (var ctx = ConnectionManager<UniConnection>.GetManager(this.ConnectionString, false, IsolationLevel.Unspecified))
{
...
}
```
An example:
1st call to TransactionManager has no IsolationLevel specified and creates a transaction.
2nd call to TransactionManager has IsolationLevel of Snapshot and must begin a new transaction since it differs from the current one (IDbTransaction.IsolationLevel)
3rd call to to TransactionManager has IsolationLevel Snapshot only increments counter for the current transaction
3rd call ends and TransactionManager counter gets decremented for the current transaction
2nd call ends and TransactionManager needs to commit the transaction
1st call ends and to TransactionManager needs to commit the transaction

Any thoughts?

**Comment by hurcane** [comment]
This seems like a viable solution to your problem. On my team, I would be concerned because new transactions are being created behind the scenes, when the developer might be expecting that the call is going to share an existing transaction. This is going to lead to extended debugging sessions before the developers realize what the framework is doing.

Given a scenario that needs to use separate transactions with different isolation levels, I would explicitly create transactions within the DataPortal methods. In my opinion, this keeps the complexity in a visible portion of the code, making life easier for future maintenance and bug hunting.

**Comment by Charles89** [comment]
That example of mixed isolation levels was merely put there, because in my opinion the framework shouldn't (or couldn't?) prevent it. In my situation I have no mixed isolation levels, but if I do have them, I would expect the TransactionManager to handle them.

In my opinion, extending the **_Transaction_**Manager to actually support mixed transactions, wouldn't make future maintenance and bug hunting harder than it is today. If you currently have transactions with mixed isolation levels, you could be in the same spot compared to having an IsolationLevel enum added to the constructor of the TransactionManager. Both are visible in the code and the working of the framework will remain as abstract as the current implementation.

**Comment by rockfordlhotka** [comment]
My concern is the same as @hurcane - what if there's an existing transaction with any (?) explicit transaction type, and you create one without an explicit transaction type? Do you just get a new one? How do you know that this happened? How would you discover that what you thought was one transaction is now multiple? The integration testing for this wouldn't be easy.
